rfcs/0182-taskcluster-yml-remote-references.md

```diff
@@ -0,0 +1,49 @@
+# RFC 182 - Allow remote references to .taskcluster.yml files processed by Taskcluster-GitHub
+* Comments: [#182](https://github.com/taskcluster/taskcluster-rfcs/pull/182)
+* Proposed by: @bhearsum
+
+# Summary
+
+Allow `.taskcluster.yml` files in GitHub repositories to outsource most of their contents to a `.taskcluster.yml` stored elsewhere.
+
+## Motivation
+
+`.taskcluster.yml` files in GitHub repositories are responsible for creating tasks (or not) in response to various GitHub events. In some cases (eg: when a decision task is used to delegate most of this to a separate tool), these files end up looking almost entirely identical, and hundreds of lines of JSON-e ends up duplicated across many different repositories. This duplication is a significant maintenance burden, and easily results in unwanted differences.
+
+Allowing a `.taskcluster.yml` to reference a file stored elsewhere will allow for deduplication of the bulk of the content in these files, resolving the concerns noted above.
+
+# Details
+
+Two new variables will be introduced to `.taskcluster.yml` files. The first, `config-from`, will allow an URL to be provided that points at a full-fledged `.taskcluster.yml` hosted elsewhere. The second, `context`, will allow for variables to be defined that will be included with the JSON-e context when the referenced `.taskcluster.yml` is rendered. Here is an example of a possible concrete `.taskcluster.yml`:
+
+```
+---
+version: 1
+config-from: https://github.com/taskcluster/taskgraph/blob/main/data/taskcluster-yml-github.yml
+context:
+  project-name: mozillavpn
+  scopes:
+    - secrets:get:project/mozillavpn/*
+```
+
+When `config-from` is present in the `.taskcluster.yml`, existing top level keys such as `policy` and `tasks` are not valid. `context` is optional, and only permitted when `config-from` is present.
+
+When Taskcluster-GitHub encounters a `.taskcluster.yml` such as the one above when processing an event for a repository, instead of directly rendering it with the context from the GitHub event, it will instead:
+1) Fetch the `.taskcluster.yml` referenced in `config-from`
+2) Combine the GitHub event context with the context provided in the repository's `.taskcluster.yml` (with GitHub event context taking priority)
```

Comment:
Will this be a deep merge? Or just wholesale replacement of top-level keys. If the latter, GH events having priority makes sense.

But if we're deep merging, project context taking priority could allow for some powerful overriding capabilities of the base config.. Though it would be pretty hacky and hard to read, so I'm not advocating for this. Just pointing out the possibility.

Reply:
Deep merge is what I was intending (...I should make that explicit).

I can certainly see that allow for overriding from the project context would allow for more possibilities...but I find it difficult to imagine a scenario where it wouldn't cause confusion or perhaps even security issues.

For example, if project context takes priority I think we could override things such as `tasks_for` or `event.pusher.email`, which I cannot imagine would be desirable.

(Maybe I'm misunderstanding what you're getting at though....)

Reply:
As I was rewriting this section for clarity, I couldn't come up with a compelling reason why deep merging would be useful, so my upcoming change declare it as shallow. (I can tweak it still, but it's a starting point for further discussion.)

Reply:
@ahal - any thoughts on the updates here?